### ISSUE TITLE
fix batch_norm in omniglot example

### ISSUE THREAD

**Issue by zou3519** [pytorch/functorch#384]
https://github.com/pytorch/functorch/blob/main/examples/maml_omniglot/maml-omniglot-transforms.py

https://github.com/pytorch/functorch/blob/main/examples/maml_omniglot/maml-omniglot-higher.py

**Comment by jgbos** [comment]
Does this now work off of the main branch?

**Comment by zou3519** [comment]
The omniglot example does not yet work in the main branch, but we can prioritize fixing it if that's something you're looking for, @jgbos.

**Comment by jgbos** [comment]
I certainly would appreciate it. I'm doing a bit of meta-learning work and these tools really simplify the workflow, but I need to do batch_norm to compare to other results.  In the mean time I'll look at group/layer_norm.

**Comment by samdow** [comment]
Hi @jgbos! Thanks for helping us prioritize this. We would love some insight into how to fix this if you have any thoughts! As a warning, this is going to be long so (1) feel free to ask any clarifying questions and (2) even only answering the first bit about how you're using BatchNorm would be extremely helpful!

**First, to check**: are you only using BatchNorm in training mode? We just want to make sure that in the cases where the input is batched but the model is not, we don't try to access running_mean or running_var (explanation below in discussion of a bigger design question). If you are, it would be awesome to see your examples so we can understand the intended behavior!

Now the bigger design question:
## Background
For some background on the problem, we're trying to prevent` vmap` from doing in-place updates on non-batched tensors because it will try to change the size of a non-batched tensor to be the batched size. For a quick code version of this with batch norm,
```
input: [B, N, C]
running_mean: [C]
running_var: [C]
vmap(torch.nn.functional.batch_norm, (0, None, None))(input, running_mean, running_var) # Error
```
this errors because `running_mean` and `running_var` will try to update to be of size `[B, C]` even though they were originally of size `[C]`, so we've only allocated `C` space in memory.

## Options
We've been considering some options on how to fix this. If you have thoughts on if you would prefer any of these (or another option), that would be super helpful!

### Option 1
We introduce something like "BatchNormNoEval" that doesn't compute a running_mean or running_var. If we try to run this after calling `nn.Module`'s `eval()` method (which is when we would use running_mean and running_var), it would error or warn that it's just using the saved mean and variance. We could introduce a helper function that swaps out BatchNorms in a model for our new batch norms, but it would still be an extra step for a user

### Option 2
We introduce an out-of-place `batch_norm`. This would return a new `running_mean` and `running_var` that could be batched if the input was batched and we could overwrite the attributes in the BatchNorm module (which lets us change the running_mean/var from `[C]` to `[B, C]` because we aren't doing it in place). The downside to this is that it would also introduce a new BatchNorm module and this module could only be run with vmap because otherwise the shape doesn't work out anymore (we've changed the running_mean/var from `[C]` to `[B, C]` which makes it invalid now)

### Option 3
Math? If people do want to use non-batched running_mean and running_var with a batched input, we can try to figure out what the behavior should be if we follow for-loop semantics so that the updated versions match the original shape. To do this, it would be great to have a use case with a predefined idea of what they want in this case so that we aren't coming up with something that disagrees with how it will be used

**Comment by jgbos** [comment]
🤯 So much to think about!  The quick answer is that I have some variants of meta-learning I'm working on, but essentially doing the exact same thing as the omniglot maml example.  So if you get the ominiglot example running, it should work for me!

As for the options, I'll have to spend some time thinking about it.

**Comment by samdow** [comment]
Thanks so much! Your usage as a data point is super helpful

Totally understand that it's a lot. As it marinates, if you have any further thoughts, definitely share! We'll keep you updated as we continue discussing this offline too

**Comment by samdow** [comment]
Closing, this is fixed and we've added a [website note](https://pytorch.org/functorch/stable/batch_norm.html) for how to fix it on other models
